### ISSUE TITLE
gin.Context is not a interface, how to extend it?

### ISSUE THREAD

**Issue by NelsonKen** [gin-gonic/gin#1123]
gin.Context is not a interface, how to extend it?

**Comment by ycavatars** [comment]
What's your scenario? I think context is the core part of gin that coordinates other components, like controls the middlewares execution, and model binding...etc. It'll make the system more complicated if you change the default behavior of such core components.

**Comment by okyantoro** [comment]
do you mean you want to share service for all request handlers globally?

**Comment by PotterDai** [comment]
Interface makes it possible to extend context, for example, you can add UserID field to context, and call context.UserID directly instead of using context.GetString("UserID").

**Comment by NarHakobyan** [comment]
any news? :)

**Comment by ww9** [comment]
Here's some input in favor of supporting custom Context in Gin. I'm not very experienced so take it with a grain of salt.

As much as I like Gin, it doesn't seem to allow for custom Context. So Instead of doing something like:

```Go
type MyContext struct {
    *gin.Context

    // Value set in my custom middleware
    User *AuthenticatedUser
}

func indexHandler (c *MyContext) {
	fmt.Print(c.User)
}
```
Advantages: 

✅ c.User is typed
✅ no typecasting from interface{}
✅ no map lookup
✅ wont compile if dev mistypes c.User
✅ is available to all handlers, per request, without extra work
✅ editor autocompletion at all times

Instead Gin forces me to do something along the lines of:

```Go
func indexHandler (c *MyContext) {
	user, ok =: c.Get("user").(*AuthenticatedUser)
	if ok {
		fmt.Print(user)
	} else {
		// handle case when "user" key was not found somehow
	}
}
```

❌ bloat
❌ typecast from interface{}
❌ map lookup 
❌ "user" key prone to mistyping*
❌ editors wont autocomplete until you finished casting it

*Note that we are not supposed to use strings as keys in stuff like that, which alleviates some of the problems (see: https://golang.org/pkg/context/#WithValue). Although there's code on Gin's frontpage using string as key (https://i.imgur.com/Kdbr9rg.png), perhaps for simplicity sake.

There are ways to mitigate some of this. One is encapsulating the typecasting logic in helper functions, typically contained inside an App struct so we don't pollute global scope:

```Go
func (app *App) User(c *gin.Context) *AuthenticatedUser {
	if user, ok =: c.Get("user").(*AuthenticatedUser); ok {
		return user
	}
	return User{} // Or perhaps nil
}
```

So now our handlers look something along the lines of:


```Go
func (app *App) indexHandler (c *MyContext) {
	fmt.Print(app.User(c))
}
```

To me this look like avoidable bloat. Specially because, from my brief evaluation of alternatives so far, gorilla mux and echo seems to support specifying custom Context:

* https://echo.labstack.com/guide/context

* https://medium.com/@nitronick600/extending-gorilla-mux-tutorial-pt-1-90d0ef3affec

I hope my point was understandable.

**Comment by ww9** [comment]
After some careful consideration I decided to go with the helper() solution described in  **Mitigating Context map problems**. Some considerations:

- Injecting dependencies (like Db connection) in handlers is done via App struct fields which is not as ideal as the more functional approach of passing them via parameters when building each handler.

- I should think about making the App struct and their handlers testable. Perhaps not all routes but I'm sure I'll want to test some routes and in those tests I'll want to pass fake dependencies such as a fake (or lightweight SQLite) database connection.

- Accessing request scoped data like authenticated user will still require writing a helper function that executes a map lookup and typecast from interface{} every time it is accessed. But since this bloat is encapsulated, I can live with that. For now.

**Comment by unlikezy** [comment]
One bad thing of gin.Context is that it's hard to cooperate with standard Context.
For example, to add user's KV to gin.Context:
```
func UserHandler(gc *gin.Context) {
        ctx := context.WithValue(gc, "user_defined_key", "user_value")
         UserFunc1(ctx)
}

func UserFunc1(ctx context.Context) {
       //do something with user_defined_key
       v, ok := ctx.Value("user_defined_key").(string)
       //.....

       //we can never get gin.Context back anyway...
}
```

so, I modify the gin.Context,  to make it more compatible with standard context:
https://github.com/unlikezy/gin

**Comment by LaysDragon** [comment]
yeah,encounter same problem here,I need to get gin.Context work with OpenTracing client packet which work with context.Context,but cannot get gin.Context back...hope I can simply get gin.Context from context.Context,at least provide some method can merge data from context.Context to gin.Context

**Comment by LaysDragon** [comment]
well fine,it looks like almost impossible to copy all value from standard context since valueCtx is private recursive structure...especially the key is private type in opentracing...I can't even use the key manually outside that package. I hope there is any way to pass regular context's Values though gin.context.

**Comment by LaysDragon** [comment]
ok I found that I can pass context though Request's ctx ,not pretty but it should work

**Comment by jarrodhroberson** [comment]
One approach that I learned when researching heterogeneous type-safe map implementations in Java was to put the method for converting the type as a method on the Key.
This would kill two birds with one stone, 1) you would have a unique type as a key and not a `string` as is recommended. 2) you have the key already since you are passing it in, just use the same instance to convert the `interface{}` to the correct type. That is basically what all the implementations in Java ended up doing.

Unfortunately that does very little good with `*gin.Context` because `.Set()` uses a `string` as the key. :-(

**Comment by Pitasi** [comment]
Just brainstormig: now that we have generics it would be nice to have gin.Context as a type constraint and implement our custom context types like

```go
type MyContext struct {
  gin.BaseContext // this itself satisfy the constraint
  MyUser *User
}
```

then I could register that type somewhere when initialising the gin.Engine and it would be passed around instead of the current gin.Context. (Not 100% sure how to do this in Go 1.18)

Of course every requests starts with a fresh context so it's my (the gin user) responsibility to fill that MyUser field somehow (i.e. in a middleware)

**Comment by jarrodhroberson** [comment]
A way to do what @Pitasi is suggesting and configure the custom stuff is to provide a `Factory` or `Provider` function would provide instances of a `gin.Context` compliant implementation. This would obviously just override a `default` context provider that just created `gin.Context` just like it does now.

`func ContextProvider(func() gin.Context) {}`

Or something similar, then we could register a custom `Context` that was a `gin.Context` with additional functionality without breaking anything.

This kind of refactoring should not break anything because it is just adding a new method and not removing anything and if the default behavior stays the same it would continue to work with existing code.

**Comment by jarrodhroberson** [comment]
> Just brainstormig: now that we have generics it would be nice to have gin.Context as a type constraint and implement our custom context types like
> 
> ```go
> type MyContext struct {
>   gin.BaseContext // this itself satisfy the constraint
>   MyUser *User
> }
> ```
> 
> then I could register that type somewhere when initialising the gin.Engine and it would be passed around instead of the current gin.Context. (Not 100% sure how to do this in Go 1.18)
> 
> Of course every requests starts with a fresh context so it's my (the gin user) responsibility to fill that MyUser field somehow (i.e. in a middleware)
> 
> btw the helper function can be a little more _generic_ now:
> 
> ```go
> func GetValue[T any](c *gin.Context, key string) T {
> 	v, ok := c.Get(key)
> 	if !ok {
> 		panic("key not found in context")
> 	}
> 	return v.(T)
> }
> ```

That is what `Producer/Consumer` pattern is for. You do not just provide the `type` you provide a `Producer[T]` that can generate a default instance for you that you can customize. Then no code really has to change and you do not have to deal with providing a customized instance anywhere "everytime".

**Comment by hamza72x** [comment]
*2023?

*almost

**Comment by YuZongYangHi** [comment]
``
func handleFunc(handler func(extender *base.ContextExtender)) func(*gin.Context) {
	return func(c *gin.Context) {
		customContext := base.ContextExtender{Context: c}
		handler(&customContext)
	}
}

func Role(c *base.ContextExtender) {
	c.List()
	c.JSON(200, "OK")
}

func NewRouter() *gin.Engine {
	router := gin.Default()

	v1 := router.Group("/openapi/v1")
	v1.Use()
	v1.Use(AuthenticationMiddleware())
	{
		v1.GET("/rbac/role", handleFunc(Role))
	}

	return router
}

``

**Comment by junderhill** [comment]
The option to provide a custom context producer would be incredibly useful.
I’ve recently been trying to use a request logger that has additional log attributes appended to it throughout the request pipeline. The issue is that with middleware we only have access to `*gin.Context` - If this could be some custom struct that is composed of `gin.Context` plus additional properties it would have solved all the issues.

**Comment by flc1125** [comment]
> 我在研究 Java 中的异构类型安全映射实现时学到的一种方法是将转换类型的方法作为 Key 上的方法。 这会用一石杀死两只鸟，1）你将有一个独特的类型作为钥匙，而不是`string`像建议的那样。 2）您已经拥有密钥，因为您正在传递它，只需使用相同的实例将其转换`interface{}`为正确的类型。这基本上就是 Java 中所有实现的最终结果。
> 
> 不幸的是，这没有什么好处，`*gin.Context`因为`.Set()`使用 a`string`作为键。 :-(



> One approach that I learned when researching heterogeneous type-safe map implementations in Java was to put the method for converting the type as a method on the Key. This would kill two birds with one stone, 1) you would have a unique type as a key and not a `string` as is recommended. 2) you have the key already since you are passing it in, just use the same instance to convert the `interface{}` to the correct type. That is basically what all the implementations in Java ended up doing.
> 
> Unfortunately that does very little good with `*gin.Context` because `.Set()` uses a `string` as the key. :-(

https://github.com/gin-gonic/gin/pull/3963 I think that should solve some of the problem.

**Comment by newcepco** [comment]
im sad it has not been addressed yet :( it would've saved us so much time
